osf/models/preprint_service.py

```diff
@@ -260,3 +323,729 @@ def _send_preprint_confirmation(self, auth):
+        """Return whether ``user`` is a contributor on this node."""
+        return user is not None and PreprintContributor.objects.filter(user=user, preprint=self).exists()
+
+    def add_contributor(self, contributor, permission=None, visible=True,
+                        send_email='default', auth=None, log=True, save=False):
+        """Add a contributor to the project.
+
+        :param User contributor: The contributor to be added
+        :param list permissions: Permissions to grant to the contributor
+        :param bool visible: PreprintContributor is visible in project dashboard
+        :param str send_email: Email preference for notifying added contributor
+        :param Auth auth: All the auth information including user, API key
+        :param bool log: Add log to self
+        :param bool save: Save after adding contributor
+        :returns: Whether contributor was added
+        """
+        MAX_RECENT_LENGTH = 15
+
+        # If user is merged into another account, use master account
+        contrib_to_add = contributor.merged_by if contributor.is_merged else contributor
+        if contrib_to_add.is_disabled:
+            raise ValidationValueError('Deactivated users cannot be added as contributors.')
+
+        if not self.is_contributor(contrib_to_add):
+
+            contributor_obj, created = PreprintContributor.objects.get_or_create(user=contrib_to_add, preprint=self)
+            contributor_obj.visible = visible
+            if not permission:
+                permission = 'write'
+            self.add_permission(contrib_to_add, permission, save=True)
+
+            contributor.save()
+
+            # Add contributor to recently added list for user
+            if auth is not None:
+                user = auth.user
+                recently_added_contributor_obj, created = RecentlyAddedContributor.objects.get_or_create(
+                    user=user,
+                    contributor=contrib_to_add
+                )
+                recently_added_contributor_obj.date_added = timezone.now()
+                recently_added_contributor_obj.save()
+                count = user.recently_added.count()
+                if count > MAX_RECENT_LENGTH:
+                    difference = count - MAX_RECENT_LENGTH
+                    for each in user.recentlyaddedcontributor_set.order_by('date_added')[:difference]:
+                        each.delete()
+            # if log:
+            #     self.add_log(
+            #         action=NodeLog.CONTRIB_ADDED,
+            #         params={
+            #             'project': self.parent_id,
+            #             'node': self._primary_key,
+            #             'contributors': [contrib_to_add._primary_key],
+            #         },
+            #         auth=auth,
+            #         save=False,
+            #     )
+            if save:
+                self.save()
+
+            if self._id:
+                project_signals.contributor_added.send(self,
```

Comment:
Does this break? The subscription code that listens for this signal assumes that the sender is a `Node`, not a `PreprintService`.